server/src/main/java/org/elasticsearch/repositories/ShardGeneration.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories;
+
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshots;
+import org.elasticsearch.snapshots.SnapshotsService;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Random;
+
+/**
+ * The generation ID of a shard, used to name the shard-level {@code index-$SHARD_GEN} file that represents a {@link
+ * BlobStoreIndexShardSnapshots} instance. Before 7.6 ({@link SnapshotsService#SHARD_GEN_IN_REPO_DATA_VERSION}) these generations were
+ * numeric, but recent versions use a UUID instead.
+ */
+public final class ShardGeneration implements Writeable, ToXContentFragment {
+
+    private final String rawGeneration;
```

Comment:
I pondered alternatives e.g. a more memory-efficient representation such as just a `byte[]`. I think we don't have too many of these things in play at once, O(1) per shard in the repository, so decided not to bother yet.

Reply:
++ the memory use of these been somewhere between trivial and small in practice even in the worst case scenarios where large repos were causing trouble, there's better spots to invest in if we wanted to improve snapshot memory consumption.